src/lib.rs

```diff
@@ -1822,172 +1815,175 @@ impl Stream {
+        match event.event_type {
+            crate::CONNECTION_EVENT_CONNECTED => {
+                let local_addr = connection.get_local_addr().unwrap().as_socket().unwrap();
+                let remote_addr = connection.get_remote_addr().unwrap().as_socket().unwrap();
+                println!("Connected({}, {})", local_addr, remote_addr);
+            }
+            crate::CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT => {
+                println!("Transport shutdown 0x{:x}", unsafe {
+                    event.payload.shutdown_initiated_by_transport.status
+                })
+            }
+            crate::CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_PEER => {
+                println!("App shutdown {}", unsafe {
+                    event.payload.shutdown_initiated_by_peer.error_code
+                })
+            }
+            crate::CONNECTION_EVENT_SHUTDOWN_COMPLETE => println!("Shutdown complete"),
+            crate::CONNECTION_EVENT_PEER_STREAM_STARTED => {
+                println!("Peer stream started");
+                unsafe {
+                    connection.set_stream_callback_handler(
+                        event.payload.peer_stream_started.stream as HQUIC,
+                        test_stream_callback,
+                        context,
+                    );
+                }
+            }
+            _ => println!("Other callback {}", event.event_type),
         }
-        _ => println!("Other callback {}", event.event_type),
-    }
-    0
-}
-
-#[allow(dead_code)] // Used in test code
-extern "C" fn test_stream_callback(
-    stream: Handle,
-    context: *mut c_void,
-    event: &StreamEvent,
-) -> u32 {
-    let connection = unsafe { &*(context as *const Connection) };
-    match event.event_type {
-        STREAM_EVENT_START_COMPLETE => println!("Stream start complete 0x{:x}", unsafe {
-            event.payload.start_complete.status
-        }),
-        STREAM_EVENT_RECEIVE => println!("Receive {} bytes", unsafe {
-            event.payload.receive.total_buffer_length
-        }),
-        STREAM_EVENT_SEND_COMPLETE => println!("Send complete"),
-        STREAM_EVENT_PEER_SEND_SHUTDOWN => println!("Peer send shutdown"),
-        STREAM_EVENT_PEER_SEND_ABORTED => println!("Peer send aborted"),
-        STREAM_EVENT_PEER_RECEIVE_ABORTED => println!("Peer receive aborted"),
-        STREAM_EVENT_SEND_SHUTDOWN_COMPLETE => println!("Peer receive aborted"),
-        STREAM_EVENT_SHUTDOWN_COMPLETE => {
-            println!("Stream shutdown complete");
-            connection.stream_close(stream);
+        0
+    }
+
+    extern "C" fn test_stream_callback(
+        stream: HQUIC,
+        _context: *mut c_void,
+        event: &StreamEvent,
+    ) -> u32 {
+        match event.event_type {
+            crate::STREAM_EVENT_START_COMPLETE => {
+                println!("Stream start complete 0x{:x}", unsafe {
+                    event.payload.start_complete.status
```

Comment:
Not specific to this PR, but why is accessing `event.payload.start_complete.status` considered _unsafe_?

Reply:
payload is a union type. Rust union is for ffi only (without metadata), and the compiler has no way to check if the access is safe. However, Rust enum has metadata embedded in it so enum is safe. I have another branch that does the union to enum conversion (in unsafe code).